java/ql/lib/semmle/code/java/security/performance/ReDoSUtil.qll

```diff
@@ -1240,3 +1240,82 @@ module ReDoSPruning<isCandidateSig/2 isCandidate> {
     result = str.suffix(str.length() - i) + str.prefix(str.length() - i)
   }
 }
+
+/**
+ * A module that describes a tree where each node has one or more accosiated characters.
+ * The root node has no accosiated character.
+ * This module is a signature used in `Concretizer`.
+ */
+signature module CharTree {
+  /** A node in the tree. */
+  class CharNode;
+
+  /** Gets the previous node in the tree from `t`. */
+  CharNode getPrev(CharNode t);
+
+  /**
+   * Holds if `n` is at the end of a tree. I.e. a node that should have a result in the `Concretizer` module.
+   * A leaf can still have children.
+   */
+  predicate isARelevantEnd(CharNode n);
+
+  /** Gets a char associated with `t`. */
+  string getChar(CharNode t);
+}
+
+/**
+ * Implements an algorithm for computing all possible strings
+ * from following a tree of nodes (as described in `CharTree`).
+ *
+ * The string is build using one big concat, where all the chars are computed first.
+ * See `concretize`.
+ */
+module Concretizer<CharTree Impl> {
+  private class Node = Impl::CharNode;
+
+  private predicate getPrev = Impl::getPrev/1;
+
+  private predicate isARelevantEnd = Impl::isARelevantEnd/1;
+
+  private predicate getChar = Impl::getChar/1;
+
+  /** Holds if `n` is on a path from the root to a leaf, and is therefore relevant for the results in `concretize`. */
+  private predicate isRelevant(Node n) {
+    isARelevantEnd(n)
+    or
+    exists(Node prev | isRelevant(prev) | n = getPrev(prev))
+  }
+
+  /** Holds if `n` is a root with no predecessors. */
+  private predicate isRoot(Node n) { not exists(getPrev(n)) }
+
+  /** Gets the distance from a root to `n`. */
+  private int nodeDist(Node n) {
+    result = 0 and isRoot(n)
+    or
+    isRelevant(n) and
+    exists(Node prev | result = nodeDist(prev) + 1 | prev = getPrev(n))
+  }
+
+  /** Holds if `n` is part of a chain that we want to compute a string for. */
+  private Node getANodeInLongChain(Node end) {
+    isARelevantEnd(end) and result = end
+    or
+    exists(Node prev | prev = getANodeInLongChain(end) | result = getPrev(prev))
+  }
+
+  pragma[noinline]
+  private string getPrefixChar(Node n, int i) {
```

Comment:
```suggestion
  /**
   * Gets the `i`th character on the path from `n` back to the root (i.e. counting backwards).
   */
  pragma[noinline]
  private string getPrefixChar(Node n, int i) {
```

Reply:
The qldoc I suggested above is based on what the predicate appears to do now.

But if we set `i = nodeDist(prev)` this would just be the `i`th character (1-based) in the string, right? (1-based because the root has no character). That seems like a more natural thing to work with than this difference thing.

Then we'd just have to remove `desc` from the concat ordering below.